src/ucs/datastruct/callbackq.c

```diff
@@ -361,6 +385,85 @@ static void ucs_callbackq_purge_slow(ucs_callbackq_t *cbq)
     priv->num_slow_elems = dst_idx;
 }
 
+/* Lock must be held */
+static unsigned ucs_callbackq_oneshot_elems_dispatch(ucs_callbackq_t *cbq)
+{
+    ucs_callbackq_priv_t *priv = ucs_callbackq_priv(cbq);
+    unsigned count             = 0;
+    ucs_callbackq_oneshot_elem_t *oneshot_elem;
+    unsigned key_idx, num_keys;
+    struct {
+        void   *key;
+        size_t count; /* How many callbacks should be called, at most */
+    } * keys;
+    ucs_hlist_head_t *hlist;
+    khiter_t khiter;
+    void *key;
+
+    num_keys = kh_size(&priv->oneshot_elems);
+    if (num_keys == 0) {
+        return 0;
+    }
+
+    keys = ucs_malloc(sizeof(*keys) * num_keys, "ucs_callbackq_keys");
+    if (keys == NULL) {
+        ucs_fatal("callbackq %p: failed to allocate oneshot key array", cbq);
+    }
+
+    key_idx = 0;
+    kh_foreach_key(&priv->oneshot_elems, key, {
+        khiter = kh_get(ucs_callbackq_oneshot_elems, &priv->oneshot_elems, key);
+        hlist  = &kh_value(&priv->oneshot_elems, khiter);
+        keys[key_idx].key   = key;
+        keys[key_idx].count = ucs_hlist_length(hlist);
+        ++key_idx;
+    })
+    ucs_assertv(key_idx == num_keys, "key_idx=%u num_keys=%u", key_idx,
+                num_keys);
+
+    key_idx = 0;
+    while (key_idx < num_keys) {
+        khiter = kh_get(ucs_callbackq_oneshot_elems, &priv->oneshot_elems,
+                        keys[key_idx].key);
+        if (khiter == kh_end(&priv->oneshot_elems)) {
+            ++key_idx; /* Not found, move to next key */
+            continue;
+        }
+
+        hlist = &kh_value(&priv->oneshot_elems, khiter);
+        if (ucs_hlist_is_empty(hlist)) {
+            kh_del(ucs_callbackq_oneshot_elems, &priv->oneshot_elems, khiter);
+            ++key_idx; /* Empty list, remove and move to next key */
+            continue;
+        }
+
+        if (keys[key_idx].count == 0) {
+            /* Should not call amy more callbacks from this key. This avoids
```

Comment:
if a callbacks adds another callback, it can result in an infinite loop

Reply:
a, you mean callback add callbacks for other keys? It will not be infinite (we can store the count for each key when start processing it), but can be quite large eventually

Reply:
when a callback add more callbacks with the same key, the list will never become empty

Reply:
but you get initial value of count (i.e. list elements), so can handle just that number of callbacks

Reply:
i meant with one loop we could save the list length before processing it and do not process more if something is added